src/execution_plan/optimizations/utilize_indices.c

```diff
@@ -19,136 +23,411 @@ AST_Operator _reverseOp(AST_Operator op) {
+			left = _filterTreeToQueryNode(filter->cond.left, sp);
+			right = _filterTreeToQueryNode(filter->cond.right, sp);
+			RediSearch_QueryNodeAddChild(node, left);
+			RediSearch_QueryNodeAddChild(node, right);
+		default:
+			assert("unexpected conditional operation");
+		}
+	} else if(filter->t == FT_N_PRED) {
+		// Make sure left hand side is variadic and right hand side is constant.
+		_normalize_filter(filter);
+		double d;
+		const char *field = filter->pred.lhs->operand.variadic.entity_prop;
+		SIValue v = filter->pred.rhs->operand.constant;
+		switch(SI_TYPE(v)) {
+		case T_STRING:
+		case T_CONSTSTRING:
+			switch(filter->pred.op) {
+			case OP_LT:    // <
+				node = RediSearch_CreateLexRangeNode(sp, field, RSLEXRANGE_NEG_INF, v.stringval, 0, 0);
+				break;
+			case OP_LE:    // <=
+				node = RediSearch_CreateLexRangeNode(sp, field, RSLEXRANGE_NEG_INF, v.stringval, 0, 1);
+				break;
+			case OP_GT:    // >
+				node = RediSearch_CreateLexRangeNode(sp, field, v.stringval, RSLECRANGE_INF, 0, 0);
+				break;
+			case OP_GE:    // >=
+				node = RediSearch_CreateLexRangeNode(sp, field, v.stringval, RSLECRANGE_INF, 1, 0);
+				break;
+			case OP_EQUAL:  // ==
+				node = RediSearch_CreateTokenNode(sp, field, v.stringval);
+				break;
+			case OP_NEQUAL: // !=
+				node = RediSearch_CreateNotNode(sp);
+				RediSearch_QueryNodeAddChild(node, RediSearch_CreateTokenNode(sp, field, v.stringval));
+				break;
+			default:
+				assert("unexpected operation");
+			}
+			break;
+
+		case T_DOUBLE:
+		case T_INT64:
+		case T_BOOL:
+			d = SI_GET_NUMERIC(v);
+			switch(filter->pred.op) {
+			case OP_LT:    // <
+				node = RediSearch_CreateNumericNode(sp, field, d, RSRANGE_NEG_INF, false, false);
+				break;
+			case OP_LE:    // <=
+				node = RediSearch_CreateNumericNode(sp, field, d, RSRANGE_NEG_INF, true, false);
+				break;
+			case OP_GT:    // >
+				node = RediSearch_CreateNumericNode(sp, field, RSRANGE_INF, d, false, false);
+				break;
+			case OP_GE:    // >=
+				node = RediSearch_CreateNumericNode(sp, field, RSRANGE_INF, d, false, true);
+				break;
+			case OP_EQUAL:  // ==
+				node = RediSearch_CreateNumericNode(sp, field, d, d, true, true);
+				break;
+			case OP_NEQUAL: // !=
+				node = RediSearch_CreateNotNode(sp);
+				RediSearch_QueryNodeAddChild(node, RediSearch_CreateNumericNode(sp, field, d, d, true, true));
+				break;
+			default:
+				assert("unexpected operation");
+			}
+			break;
+		default:
+			assert("unexpected value type");
+		}
+	} else {
+		assert("unknow filter tree node type");
+	}
+	return node;
+}
+
+//------------------------------------------------------------------------------
+
+/* Tests to see if given filter tree is a simple predicate
+ * e.g. n.v = 2
+ * one side is variadic while the other side is constant. */
+bool _simple_predicates(const FT_FilterNode *filter) {
+	if(filter->t == FT_N_PRED) {
+		if(filter->pred.lhs->type == AR_EXP_OP || filter->pred.rhs->type == AR_EXP_OP) {
+			return false;
```

Comment:
Negative numbers fail this condition:
```
$ ge social "MATCH (a:person) WHERE a.age > -1 RETURN a"
1) "Results"
2) "    Project"
3) "        Filter"
4) "            Node By Label Scan | (a:person)"
```
Since they are multiplication operation trees. This should be either fixed or noted in a TODO, I think.

Reply:
I'll introduce `ReduceToScalar` function to arithmetic expression, such that when we construct an expression tree when possible we'll a node such as minus(exp) will be reduced to a single const scalar if all nodes within `exp` aren't of type variadic not aggregate.